bluemira/equilibria/optimisation/constraints.py

```diff
@@ -445,6 +449,71 @@ def f_constraint(self) -> ConstraintFunction:
         return f_constraint
 
 
+class SphericalHarmonicsConstraint(MagneticConstraint):
+    """
+    Constraint function to constrain spherical harmonics starting from initial
+    coil currents and associated core plasma.
+
+    Parameters
+    ----------
+    r_t:
+        Typical length scale of the problem (e.g. radius at outer midplane)
+    ref_coil_psi_amplitudes:
+        Initial harmonic amplitudes obtained from desired core plasma
+    max_degree:
+        Maximum degree of spherical harmonics desired to constrain.
+    """
+
+    def __init__(
+        self,
+        r_t: float,
+        ref_coil_psi_amplitudes: npt.NDArray,
+        max_degree: int,
+        target_value: float = 0.0,
+        weights: Union[float, npt.NDArray] = 1.0,
+        tolerance: Union[float, npt.NDArray] = 1e-6,
+        f_constraint: Type[ConstraintFunction] = AxBConstraint,
+        constraint_type: str = "inequality",
+    ):
+        super().__init__(
+            target_value,
+            weights,
+            tolerance=tolerance,
+            f_constraint=f_constraint,
+            constraint_type=constraint_type,
+        )
+        self.max_degree = max_degree
+        self.r_t = r_t
+        self.ref_coil_psi_amplitudes = ref_coil_psi_amplitudes
+
+    def control_response(self, coilset: CoilSet) -> npt.NDArray:
+        """
+        Calculate control response of a CoilSet to the constraint.
+        """
+        return coil_harmonic_amplitude_matrix(coilset, self.max_degree, self.r_t)
+
+    def evaluate(self, _: Equilibrium) -> npt.NDArray:
```

Comment:
Note: Check if the prepare method has a bug in it where it's setting b_vec = self.target_value - self.evaluate()

Reply:
for clarity this could be a bug that has existed for a while (before the optimisation refactor). Just needs checking